json_serializable/lib/src/json_key_utils.dart

```diff
@@ -94,36 +94,50 @@ JsonKey _from(FieldElement element, JsonSerializable classAnnotation) {
         'Please rerun your build with `--verbose` and file an issue.');
   }
 
-  final defaultValueObject = obj.getField('defaultValue');
+  Object _annotationValue(String fieldName, bool mustBeEnum) {
+    final annotationValue = obj.getField(fieldName);
 
-  Object defaultValueLiteral;
-
-  final enumFields = iterateEnumFields(defaultValueObject.type);
-  if (enumFields != null) {
-    final enumValueNames =
-        enumFields.map((p) => p.name).toList(growable: false);
-
-    final enumValueName = enumValueForDartObject<String>(
-        defaultValueObject, enumValueNames, (n) => n);
-
-    defaultValueLiteral = '${defaultValueObject.type.name}.$enumValueName';
-  } else {
-    defaultValueLiteral = _getLiteral(defaultValueObject, []);
-    if (defaultValueLiteral != null) {
-      defaultValueLiteral = jsonLiteralAsDart(defaultValueLiteral);
+    final enumFields = iterateEnumFields(annotationValue.type);
+    if (enumFields != null) {
+      if (mustBeEnum && !isEnum(element.type)) {
+        throwUnsupported(
+          element,
+          '`$fieldName` can only be set on fields of type enum.',
+        );
+      }
+      final enumValueNames =
+          enumFields.map((p) => p.name).toList(growable: false);
+
+      final enumValueName = enumValueForDartObject<String>(
+          annotationValue, enumValueNames, (n) => n);
+
+      return '${annotationValue.type.name}.$enumValueName';
+    } else {
+      final defaultValueLiteral = _getLiteral(annotationValue, []);
+      if (defaultValueLiteral != null) {
```

Comment:
Consider a short circuit to save 1 level of nesting - we get up to 3 levels here.

```dart
final defaultValueLiteral = _getLiteral(annotationValue, []);
if (defaultValueLiteral == null) return null;
if (mustBeEnum) {
  ...
```

Reply:
ack